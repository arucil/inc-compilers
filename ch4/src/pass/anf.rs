use ast::{Exp, IdxVar, Program};
use support::Range;

pub fn anf(prog: Program<IdxVar>) -> Program<IdxVar> {
  let mut counter = 0;
  Program {
    body: prog
      .body
      .into_iter()
      .map(|(range, exp)| (range, anf_exp(range, exp, &mut counter)))
      .collect(),
  }
}

fn anf_exp(range: Range, exp: Exp<IdxVar>, counter: &mut usize) -> Exp<IdxVar> {
  match exp {
    Exp::Let { var, init, body } => Exp::Let {
      var,
      init: box (init.0, anf_exp(init.0, init.1, counter)),
      body: box (body.0, anf_exp(body.0, body.1, counter)),
    },
    Exp::Prim { op, args } => {
      let mut tmps = vec![];
      let args = args
        .into_iter()
        .map(|(range, arg)| (range, atom_exp(range, arg, &mut tmps, counter)))
        .collect();
      tmps
        .into_iter()
        .rfold(
          (range, Exp::Prim { op, args }),
          |(body_range, body), (range, var, init)| {
            (
              body_range,
              Exp::Let {
                var: (range, var),
                init: box (range, init),
                body: box (body_range, body),
              },
            )
          },
        )
        .1
    }
    Exp::If { cond, conseq, alt } => Exp::If {
      cond: box (cond.0, anf_exp(cond.0, cond.1, counter)),
      conseq: box (conseq.0, anf_exp(conseq.0, conseq.1, counter)),
      alt: box (alt.0, anf_exp(alt.0, alt.1, counter)),
    },
    Exp::Set { var, exp } => Exp::Set {
      var,
      exp: box (exp.0, anf_exp(exp.0, exp.1, counter)),
    },
    Exp::Begin { seq, last } => Exp::Begin {
      seq: seq
        .into_iter()
        .map(|exp| (exp.0, anf_exp(exp.0, exp.1, counter)))
        .collect(),
      last: box (last.0, anf_exp(last.0, last.1, counter)),
    },
    Exp::While { cond, body } => Exp::While {
      cond: box (cond.0, anf_exp(cond.0, cond.1, counter)),
      body: box (body.0, anf_exp(body.0, body.1, counter)),
    },
    Exp::Print { val, ty } => Exp::Print {
      val: box (val.0, anf_exp(val.0, val.1, counter)),
      ty,
    },
    exp => exp,
  }
}

fn atom_exp(
  range: Range,
  exp: Exp<IdxVar>,
  tmps: &mut Vec<(Range, IdxVar, Exp<IdxVar>)>,
  counter: &mut usize,
) -> Exp<IdxVar> {
  match exp {
    Exp::Prim { op, args } => {
      let args = args
        .into_iter()
        .map(|(range, arg)| (range, atom_exp(range, arg, tmps, counter)))
        .collect();
      let exp = Exp::Prim { op, args };
      assign_var(range, exp, tmps, counter)
    }
    Exp::Let { .. }
    | Exp::If { .. }
    | Exp::Set { .. }
    | Exp::Begin { .. }
    | Exp::While { .. } => {
      let exp = anf_exp(range, exp, counter);
      assign_var(range, exp, tmps, counter)
    }
    exp => exp,
  }
}

fn assign_var(
  range: Range,
  exp: Exp<IdxVar>,
  tmps: &mut Vec<(Range, IdxVar, Exp<IdxVar>)>,
  counter: &mut usize,
) -> Exp<IdxVar> {
  let tmp = IdxVar {
    name: "tmp".to_owned(),
    index: *counter,
  };
  *counter += 1;
  tmps.push((range, tmp.clone(), exp));
  Exp::Var(tmp)
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn init_with_var() {
    let prog = parse(r#"(let ([a 42]) (let ([b a]) b))"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_in_init() {
    let prog = parse(
      r#"(let ([a (+ (let ([x (read)] [y (- x)]) (+ x (- y))) 7)]) (- a))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_form() {
    let prog = parse(
      r#"(+ (- 3 2) (if (not (eq? 2 3)) (- (read)) (> (read) (not #t))))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn begin() {
    let prog = parse(
      r#"
(let ([x0 10])
  (let ([y1 0])
    (+ (+ (begin (set! y1 (read)) x0)
       (begin (set! x0 (read)) y1))
    x0)))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
