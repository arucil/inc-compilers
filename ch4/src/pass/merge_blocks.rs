use asm::{Block, Instr, Label, Program};
use indexmap::IndexMap;

pub fn merge_blocks<INFO>(prog: Program<INFO>) -> Program<INFO> {
  let mut refs = IndexMap::<Label, usize>::new();
  let mut blocks = IndexMap::<Label, Block>::new();
  for (label, block) in prog.blocks {
    if let Instr::JmpLabel(label) = block.code.last().unwrap() {
      *refs.entry(*label).or_default() += 1;
      if block.code.len() > 1 {
        if let Instr::JumpIf { label, .. } = &block.code[block.code.len() - 2] {
          *refs.entry(*label).or_default() += 1;
        }
      }
    }
    blocks.insert(label, block);
  }

  let mut worklist = vec![Label::EntryPoint];
  let mut new_blocks = vec![];
  while let Some(mut label) = worklist.pop() {
    let mut block = if let Some(block) = blocks.remove(&label) {
      block
    } else {
      continue;
    };
    while let Instr::JmpLabel(next_label) = block.code.last().unwrap() {
      if block.code.len() > 1 {
        if let Instr::JumpIf { label, .. } = &block.code[block.code.len() - 2] {
          worklist.push(*label);
        }
      }

      let next_label = *next_label;
      if refs[&next_label] == 1 {
        let mut next_block = blocks.remove(&next_label).unwrap();
        block.code.pop();
        block.code.append(&mut next_block.code);
      } else if let Some(next_block) = blocks.remove(&next_label) {
        block.code.pop();
        new_blocks.push((label, block));
        label = next_label;
        block = next_block;
      } else {
        break;
      }
    }
    new_blocks.push((label, block));
  }
  Program {
    blocks: new_blocks,
    ..prog
  }
}
