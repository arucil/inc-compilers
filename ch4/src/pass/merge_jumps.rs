use asm::{Program, Label, Instr, Block};
use indexmap::IndexMap;

pub fn merge_jumps<INFO>(prog: Program<INFO>) -> Program<INFO> {
  let mut refs = IndexMap::<Label, usize>::new();
  let mut blocks = IndexMap::<Label, Block>::new();
  for (label, block) in prog.blocks {
    if let Instr::Jmp(label) = block.code.last().unwrap() {
      *refs.entry(label.clone()).or_default() += 1;
      if block.code.len() > 1 {
        if let Instr::Jmp(label) = &block.code[block.code.len() - 2] {
          *refs.entry(label.clone()).or_default() += 1;
        }
      }
    }
    blocks.insert(label, block);
  }
  let mut label = Label::EntryPoint;
  let mut block = blocks.remove(&label).unwrap();
  let mut new_blocks = vec![];
  while let Instr::Jmp(next_label) = block.code.last().unwrap() {
    let next_label = *next_label;
    if refs[&next_label] == 1 {
      let mut next_block = blocks.remove(&next_label).unwrap();
      block.code.pop();
      block.code.append(&mut next_block.code);
    } else {
      block.code.pop();
      new_blocks.push((label, block));
      label = next_label;
      block = blocks.remove(&label).unwrap();
    }
  }
  new_blocks.push((label, block));
  new_blocks.extend(blocks.into_iter());
  Program {
    info: prog.info,
    blocks: new_blocks,
  }
}