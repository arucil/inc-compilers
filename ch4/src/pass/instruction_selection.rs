use super::explicate_control::CInfo;
use asm::Program;
use ast::{IdxVar, Type};
use ch2::pass::instruction_selection::{CodeGen, Locals};
use control::*;
use indexmap::IndexMap;
use std::fmt::{self, Display, Formatter};

#[derive(Default)]
pub struct Info {
  pub locals: IndexMap<IdxVar, Type>,
}

impl Display for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

impl From<CInfo> for Info {
  fn from(info: CInfo) -> Self {
    Self {
      locals: info.locals,
    }
  }
}

impl Locals for Info {
  fn add_local(&mut self, local: IdxVar, ty: Type) {
    self.locals.insert(local, ty);
  }
}

pub fn select_instruction(
  prog: CProgram<CInfo>,
  use_heap: bool,
) -> Program<Info, IdxVar> {
  let types = prog.types;
  let mut codegen = CodeGen::new(&types, use_heap);
  let funs = prog
    .funs
    .into_iter()
    .map(|fun| codegen.gen_fun(fun))
    .collect();
  let blocks = codegen.gen_body(prog.body);
  let result = codegen.finish();
  Program {
    info: Info {
      locals: prog.info.locals,
    },
    blocks,
    constants: result.constants,
    externs: result.externs,
    funs,
    types,
  }
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use ast::*;
  use ch2::pass::remove_complex_operands;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = select_instruction(prog, false);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_if() {
    let prog = ast::parse(
      r#"
(let ([x (read)]
      [y (read)])
  (if (if (< x 1)
        (eq? x 0)
        (eq? x 2))
    (+ y 2)
    (+ y 10)))
    "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = select_instruction(prog, false);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_in_init() {
    let prog = ast::parse(
      r#"(let ([x (if (>= (read) 3) 10 77)]) (if (not (eq? x 10)) 41 2))"#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = select_instruction(prog, false);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn complex_if() {
    let prog = ast::parse(
      r#"
(let ([x (read)])
  (if (> x 100)
    (let ([y (read)])
      (if (or (and (>= x y) (< (- x y) 10))
              (and (< x y) (< (- y x) 10)))
        -1
        (- y)))
    (if (and (> x 40) (< x 60))
      5000
      x)))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = select_instruction(prog, false);
    assert_snapshot!(result.to_string_pretty());
  }
}
