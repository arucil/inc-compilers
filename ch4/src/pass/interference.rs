#[cfg(test)]
mod tests {
  use super::super::*;
  use asm::{Label, Program};
  use ast::IdxVar;
  use ch2::pass::instruction_selection::Info as OldOldInfo;
  use ch3::location_set::LocationSet;
  use ch3::pass::interference;
  use indexmap::indexset;
  use insta::assert_snapshot;
  use maplit::hashmap;

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      mov v, i
      mov w, 42
      mov x, v
      add x, 7
      mov y, x
      mov z, x
      add z, w
      mov t, y
      neg t
      mov rax, z
      add rax, t
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("v"),
          IdxVar::new("i"),
          IdxVar::new("w"),
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("t"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call foo, 3
      add w, rax
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("w"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn mov_same_variables() {
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      mov t, x
      add t, y
      mov z, t
      add z, w
    "#,
    );
    let label_live = hashmap! {};
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("t"),
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("w"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn if_in_init() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov tmp.0, rax
    cmp tmp.0, 3
    jge block3
    jmp block4
block0:
    mov rax, 2
    jmp conclusion
block1:
    mov rax, 41
    jmp conclusion
block2:
    cmp x.0, 10
    sete al
    movzx tmp.1, al
    cmp tmp.1, 0
    je block1
    jmp block0
block3:
    mov x.0, 10
    jmp block2
block4:
    mov x.0, 77
    jmp block2
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("tmp.0"),
          IdxVar::new("x.0"),
          IdxVar::new("tmp.1"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn nested_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        call read_int
        mov x, rax
        call read_int
        mov y, rax
        cmp x, 1
        jl block2
        jmp block3
      block0:
        mov rax, y
        add rax, 2
        jmp conclusion
      block1:
        mov rax, z
        add rax, 10
        jmp conclusion
      block2:
        cmp x, 0
        je block0
        jmp block1
      block3:
        cmp x, 2
        je block0
        jmp block1
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn complex_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov x.0, rax
    cmp x.0, 100
    jg block8
    jmp block9
block0:
    mov rax, -1
    jmp conclusion
block1:
    mov rax, y.1
    neg rax
    jmp conclusion
block2:
    mov tmp.1, y.1
    sub tmp.1, x.0
    cmp tmp.1, 10
    jl block0
    jmp block1
block3:
    cmp x.0, y.1
    jl block2
    jmp block1
block4:
    mov tmp.0, x.0
    sub tmp.0, y.1
    cmp tmp.0, 10
    jl block0
    jmp block3
block5:
    mov rax, 5000
    jmp conclusion
block6:
    mov rax, x.0
    jmp conclusion
block7:
    cmp x.0, 60
    jl block5
    jmp block6
block8:
    call read_int
    mov y.1, rax
    cmp x.0, y.1
    jge block4
    jmp block3
block9:
    cmp x.0, 40
    jg block7
    jmp block6
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("x.0"),
          IdxVar::new("y.1"),
          IdxVar::new("tmp.1"),
          IdxVar::new("tmp.0"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }
}
