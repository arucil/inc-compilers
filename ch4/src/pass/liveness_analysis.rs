use asm::{Block, Instr, Label, Program};
use ast::IdxVar;
use ch2::pass::instruction_selection::Info as OldInfo;
use ch3::location_set::{LocationSet, VarStore};
use ch3::pass::liveness_analysis::{AnalysisState, Info};
use petgraph::algo::toposort;
use petgraph::graph::NodeIndex;
use petgraph::Graph;
use std::collections::HashMap;

/// `label_live` is a map from labels to sets of live locations before the first
/// instruction of the blocks.
pub fn analyze_liveness(
  prog: Program<OldInfo, IdxVar>,
  mut label_live: HashMap<Label, LocationSet>,
) -> Program<Info, IdxVar> {
  let mut var_store = VarStore::new();
  for var in &prog.info.locals {
    var_store.insert(var.clone());
  }

  let state = AnalysisState {
    var_store: &var_store,
  };

  let live = sort_blocks(&prog.blocks)
    .map(|(label, block)| {
      let live = state.block_liveness(block, &label_live);
      label_live.insert(label, live[0].clone());
      (label, live)
    })
    .collect();

  Program {
    info: Info {
      locals: prog.info.locals,
      live,
      var_store,
    },
    constants: prog.constants,
    blocks: prog.blocks,
  }
}

fn sort_blocks(
  blocks: &[(Label, Block<IdxVar>)],
) -> impl Iterator<Item = (Label, &Block<IdxVar>)> {
  let mut graph = Graph::new();
  let mut nodes = HashMap::<Label, NodeIndex>::new();
  for (label, block) in blocks {
    let ix = graph.add_node((*label, block));
    nodes.insert(*label, ix);
  }

  for node in nodes.values() {
    let last = graph[*node].1.code.last().unwrap().clone();
    if let Instr::Jmp(label) = last {
      if let Some(&node1) = nodes.get(&label) {
        graph.add_edge(*node, node1, ());
      }
      let code = &graph[*node].1.code;
      if code.len() > 1 {
        if let Instr::JumpIf { label, .. } = code[code.len() - 2].clone() {
          graph.add_edge(*node, nodes[&label], ());
        }
      }
    }
  }

  toposort(&graph, None)
    .unwrap()
    .into_iter()
    .rev()
    .map(move |node| graph[node])
}

#[cfg(test)]
mod tests {
  use super::*;
  use asm::Label;
  use indexmap::indexset;
  use insta::assert_snapshot;
  use maplit::hashmap;

  trait ShowLiveness {
    fn show(&self) -> String;
  }

  impl ShowLiveness for Program<Info, IdxVar> {
    fn show(&self) -> String {
      let mut buf = String::new();
      for (label, block) in &self.blocks {
        buf += &format!("{:?}:\n", label);
        let live = &self.info.live[label];
        for (i, l) in live.iter().enumerate().take(block.code.len()) {
          buf += "                    ";
          l.write(&mut buf, &self.info.var_store).unwrap();
          buf += "\n";
          buf += &format!("    {:?}", block.code[i]);
          buf += "\n";
        }
        buf += "                    ";
        live
          .last()
          .unwrap()
          .write(&mut buf, &self.info.var_store)
          .unwrap();
        buf += "\n\n";
      }
      buf
    }
  }

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        mov v, 1
        mov w, 42
        mov x, v
        add x, 7
        mov y, x
        mov z, x
        add z, w
        mov t, y
        neg t
        mov rax, z
        add rax, t
        jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: indexset! {
          IdxVar::new("v"),
          IdxVar::new("w"),
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("t"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn push_pop() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        push x
        mov w, rbx
        pop rbx
        add x, w
        jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("w"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        pop rdi
        pop rsi
        push x
        mov w, rbx
        call foo, 3
        add w, rax
        jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("w"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn epilogue() {
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        mov rsp, rbp
        pop rbp
        call print_int
        call print_newline
        mov rax, 60
        mov rdi, 0
        syscall
    "#,
    );
    let label_live = HashMap::new();
    let prog = Program {
      info: OldInfo {
        locals: indexset! {},
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn nested_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        call read_int
        mov x, rax
        call read_int
        mov y, rax
        cmp x, 1
        jl block2
        jmp block3
      block0:
        mov rax, y
        add rax, 2
        jmp conclusion
      block1:
        mov rax, z
        add rax, 10
        jmp conclusion
      block2:
        cmp x, 0
        je block0
        jmp block1
      block3:
        cmp x, 2
        je block0
        jmp block1
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn if_in_init() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov tmp.0, rax
    cmp tmp.0, 3
    jge block3
    jmp block4
block0:
    mov rax, 2
    jmp conclusion
block1:
    mov rax, 41
    jmp conclusion
block2:
    cmp x.0, 10
    sete al
    movzx tmp.1, al
    cmp tmp.1, 0
    je block1
    jmp block0
block3:
    mov x.0, 10
    jmp block2
block4:
    mov x.0, 77
    jmp block2

    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: indexset! {
          IdxVar::new("tmp.0"),
          IdxVar::new("x.0"),
          IdxVar::new("tmp.1"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn complex_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov x.0, rax
    cmp x.0, 100
    jg block8
    jmp block9
block0:
    mov rax, -1
    jmp conclusion
block1:
    mov rax, y.1
    neg rax
    jmp conclusion
block2:
    mov tmp.1, y.1
    sub tmp.1, x.0
    cmp tmp.1, 10
    jl block0
    jmp block1
block3:
    cmp x.0, y.1
    jl block2
    jmp block1
block4:
    mov tmp.0, x.0
    sub tmp.0, y.1
    cmp tmp.0, 10
    jl block0
    jmp block3
block5:
    mov rax, 5000
    jmp conclusion
block6:
    mov rax, x.0
    jmp conclusion
block7:
    cmp x.0, 60
    jl block5
    jmp block6
block8:
    call read_int
    mov y.1, rax
    cmp x.0, y.1
    jge block4
    jmp block3
block9:
    cmp x.0, 40
    jg block7
    jmp block6
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: indexset! {
          IdxVar::new("x.0"),
          IdxVar::new("y.1"),
          IdxVar::new("tmp.1"),
          IdxVar::new("tmp.0"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }
}
