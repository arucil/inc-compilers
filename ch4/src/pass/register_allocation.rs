use super::interference::Info as OldInfo;
use asm::{Block, Program, Reg};
use ast::{IdxVar, Type};
use ch3::pass::register_allocation::{
  gen_assign_instr_registers, RegisterAlloc,
};
use indexmap::{IndexMap, IndexSet};
use std::collections::HashMap;
use std::fmt::{self, Formatter, Display};

pub struct Info {
  pub locals: IndexMap<IdxVar, Type>,
  /// in bytes
  pub stack_space: usize,
  pub used_callee_saved_regs: IndexSet<Reg>,
}

impl Display for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}", self.locals)?;
    write!(f, "used_callee_saved_regs: ",)?;
    let mut comma = false;
    for reg in &self.used_callee_saved_regs {
      if comma {
        write!(f, ", ")?;
      }
      comma = true;
      write!(f, "{}", reg)?;
    }
    writeln!(f, "\nstack_space: {} bytes", self.stack_space)
  }
}

pub fn allocate_registers(
  prog: Program<OldInfo, IdxVar>,
  available_regs: &[Reg],
) -> Program<Info> {
  assert!(prog.funs.is_empty());

  let (info, blocks) = alloc_body_regs(prog.info, prog.blocks, available_regs);

  Program {
    info,
    blocks,
    funs: vec![],
    ..prog
  }
}

fn alloc_body_regs(
  info: OldInfo,
  blocks: Vec<Block<IdxVar>>,
  available_regs: &[Reg],
) -> (Info, Vec<Block>) {
  let mut num_locals = 0;
  let new_blocks;
  let used_callee_saved_regs;

  {
    let reg_colors: HashMap<_, _> = available_regs
      .iter()
      .enumerate()
      .map(|(i, &reg)| (reg, i as _))
      .collect();

    let mut alloc = RegisterAlloc {
      num_locals: info.locals.len(),
      conflicts: &info.conflicts,
      moves: &info.moves,
      reg_colors: &reg_colors,
      available_regs,
      used_callee_saved_regs: IndexSet::new(),
      assign_instr_registers: gen_assign_instr_registers(
        &info.var_store,
        available_regs,
        &mut num_locals,
      ),
    };

    new_blocks = blocks
      .into_iter()
      .map(|block| alloc.allocate_block_registers(block))
      .collect();

    used_callee_saved_regs = alloc.used_callee_saved_regs;
  }

  (
    Info {
      locals: info.locals,
      stack_space: num_locals * 8,
      used_callee_saved_regs,
    },
    new_blocks,
  )
}
