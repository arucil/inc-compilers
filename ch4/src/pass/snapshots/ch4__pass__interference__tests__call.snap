---
source: ch4/src/pass/interference.rs
expression: "format!(\"{:?}\", result.info)"
---
graph {
    0 [ label = "x.0" ]
    1 [ label = "w.0" ]
    2 [ label = "rdi" ]
    3 [ label = "rsp" ]
    4 [ label = "rbx" ]
    5 [ label = "rdx" ]
    6 [ label = "rsi" ]
    7 [ label = "rax" ]
    8 [ label = "rcx" ]
    9 [ label = "r8" ]
    10 [ label = "r9" ]
    11 [ label = "r10" ]
    12 [ label = "r11" ]
    2 -- 3 [ ]
    2 -- 4 [ ]
    2 -- 5 [ ]
    2 -- 0 [ ]
    6 -- 3 [ ]
    6 -- 4 [ ]
    6 -- 5 [ ]
    6 -- 2 [ ]
    6 -- 0 [ ]
    1 -- 3 [ ]
    1 -- 5 [ ]
    1 -- 6 [ ]
    1 -- 2 [ ]
    7 -- 3 [ ]
    7 -- 1 [ ]
    8 -- 3 [ ]
    8 -- 7 [ ]
    8 -- 1 [ ]
    5 -- 3 [ ]
    5 -- 7 [ ]
    6 -- 7 [ ]
    2 -- 7 [ ]
    9 -- 3 [ ]
    9 -- 7 [ ]
    9 -- 1 [ ]
    10 -- 3 [ ]
    10 -- 7 [ ]
    10 -- 1 [ ]
    11 -- 3 [ ]
    11 -- 7 [ ]
    11 -- 1 [ ]
    12 -- 3 [ ]
    12 -- 7 [ ]
    12 -- 1 [ ]
}

